Multi-threaded Asset Loading + More

Honeybunch opened the pull request:
Re-writing the rendering systems to load assets asynchronously.

Also want to tackle the way that meshes are constructed and fed to the renderer to make it easier to implement more diverse types of drawing. Most importantly being drawing debug physics meshes for Jolt

Honeybunch committed:
Moving around some ideas for a new texture system

Honeybunch commented:
Starting with the texture system as it has the fewest direct dependencies. This is more about loading but I also want to see if I can maybe come up with a tighter collection of systems that do the same job as the current "mega" texture system.

If this pattern holds it should just be a matter of refactoring the rest of the rendering systems to match. Then the bigger challenge is making the mesh rendering systems more flexible. Which I think I have some ideas for.

Honeybunch committed:
Getting the first phase of async texture loading working

Honeybunch committed:
Setting up texture descriptor pool for texture system 2

Honeybunch committed:
Making progress at getting the new texture system in and working. Replaced usage in the view system and it works. Next we replace the old texture system entirely.

Honeybunch committed:
Replacing the old texture system

Honeybunch committed:
Fixing log to actually print

Honeybunch committed:
Vulkan debugging breakpoint

Honeybunch committed:
Adding missing config.h.in changes

Honeybunch committed:
Lots of little texture load fixes

Honeybunch committed:
Fixing sync problems

Honeybunch committed:
memory mapped io for faster asset loading

Honeybunch committed:
Adding the concept of a thread local allocator

Honeybunch committed:
Fixing build for posix platforms

Honeybunch committed:
Texture loading no longer blocks the main thread

Honeybunch committed:
Fixing silly error

Honeybunch committed:
Another attempt to fix build

Honeybunch committed:
I swear I can fix this

Honeybunch committed:
Refactoring queues

Honeybunch committed:
Some deduplication

Honeybunch committed:
Sketching out the new material system and a couple other quality of life and bug fixes

Honeybunch committed:
More sketching async material loads

Honeybunch committed:
Loading materials that load textures

Honeybunch committed:
Making progress on figuring out how to get materials to chain load textures

Honeybunch committed:
Materials go through the new loading process

Honeybunch committed:
_Thread_local to thread_local

Honeybunch committed:
A bit of renaming

Honeybunch committed:
Fixing some material sizing issues

Honeybunch committed:
Progress. Just need to fix bistro

Honeybunch committed:
Fixing materials not loading

Honeybunch committed:
Some misc fixups

Honeybunch committed:
Fixing static windows build

Honeybunch committed:
allocator.h to tb_allocator.h

Honeybunch committed:
Renaming and trimming files

Honeybunch committed:
Sketching out the new mesh system

Honeybunch committed:
Some mesh systems

Honeybunch committed:
tb_engine_config

dislogical commented on source/tb_mesh_system2.c:
```diff
@@ -0,0 +1,181 @@
+  mesh_ent = ecs_new_entity(ecs, 0);
+  ecs_set_name(ecs, mesh_ent, name);
+
+  // It is a child of the texture system context singleton
```
Copy-pasta comment typo

Honeybunch committed:
New mesh descriptor update system

Honeybunch committed:
Sketching mesh load system and fixing string dependency in material->texture loading

Honeybunch committed:
Fixing system not running

Honeybunch committed:
Sketching out some ideas for scene loading as well as getting async mesh loading to get data into driver owned memory

Honeybunch committed:
Handling submesh loading

Honeybunch committed:
Refactoring cmake presets to be a bit more flexible

Honeybunch committed:
Fixing presets

Honeybunch committed:
New loading strategy and descriptor write flow

The new mesh system is invoked which is sadly very slow due to descriptor pool thrashing during load. The solution will be to only resize the descriptor pool when a scene has finished parsing all of its resources

Honeybunch committed:
Textures now load in phases to reduce descriptor thrash

Honeybunch committed:
Similar phasing strategy for material loading

Honeybunch committed:
Scene2 drives mesh loading

Honeybunch committed:
The new mesh system is plumbed for rendering but meshes aren't drawn properly yet

Honeybunch committed:
Fixing problem of how to identify meshes across multiple sources

Honeybunch committed:
Fixing failure to deduplicate material entities

Honeybunch committed:
Pinned tasks are now a way for systems to exfiltrate work outside of a readonly system

Honeybunch committed:
Removing more of the old mesh asset loading systems

Honeybunch committed:
Reorganizing a bit of the water systems

Honeybunch committed:
Meshes render again though with some problems

Honeybunch committed:
Some instrumentation and profiling fixes

Honeybunch committed:
Finally figured out why we'd get validation failures with tracy

Honeybunch committed:
Fixing meshes

Honeybunch committed:
Fixing obvious warning

Honeybunch committed:
Fixing a small issue in FINAL builds

Honeybunch committed:
Rough but functional improvement to gltf loading

Honeybunch committed:
Fixing build

Honeybunch committed:
Major progress on async scene loading. Still working on getting entities loading correctly

Honeybunch committed:
Fixed rendering but parents are screwed

Honeybunch committed:
Scene waits for entities to have components report readiness

Honeybunch committed:
Fixing parenting issues

Honeybunch committed:
Making sure entire scene gets parsed properly

Honeybunch committed:
Removing log and fixing crash with bistro

Honeybunch committed:
Improving scene loading. Can't load bistro yet but making progress

Honeybunch committed:
Progress on making sure loading completes. Smaller systems like the ocean test work but bistro isn't there yet

Honeybunch committed:
More loading UI and making sure not to overrun descriptor pool

Honeybunch committed:
Adding a progress bar for my own sake

Honeybunch committed:
Removing an unused variable and making sure that entities popped off the queue are loaded

Honeybunch committed:
Fixing some loading problems. GPU timeout remains

Honeybunch committed:
Sketching out descriptor buffer structure and API

Honeybunch committed:
Initializing descriptor buffer in the texture system

Honeybunch committed:
Making sure the descriptor buffer can be replaced inline

Honeybunch committed:
Cleaning up some of the buffer address access pattern

Honeybunch committed:
Migrating the render object system to descriptor buffers

Honeybunch committed:
Working on getting the descriptor buffer up to meeting the needs of the view system

Honeybunch committed:
Fixing build

Honeybunch committed:
More progress on descriptor buffers

Honeybunch committed:
Migrating mesh systems to descriptor buffers

Honeybunch committed:
Descriptor Buffer code path compiles + some refactoring

Honeybunch committed:
Validation layers stopped screaming but loading never finishes

Honeybunch committed:
Loads and runs but doesn't render yet

Honeybunch committed:
Fixing silly issue with buffer indices not being set properly

Honeybunch committed:
Fixing sky

Honeybunch committed:
Defining TbShader also sketching out what an update to the sky system would look like

Honeybunch committed:
An attempt to get the sky system to work with descriptor buffers that doesn't work but doesn't crash

Honeybunch committed:
Fixed the sky Descriptor Buffer not working

Honeybunch committed:
Small fix to make sure that the mesh and material systems don't try to write more descriptors than available

Honeybunch committed:
Some cleanup and improvements to reduce crashiness

Honeybunch committed:
Some extra cleanup

Honeybunch committed:
Working on a better descriptor access pattern

Honeybunch committed:
Updating material descriptor approach

Honeybunch committed:
Updating mesh descriptor system. Doesn't work lmao

Honeybunch committed:
Fixing the dyn desc pool not writing the correct resources. Still have a GPU crash

Honeybunch committed:
Making some of the descriptor API a bit more ergonomic to rule out crash sources

Honeybunch committed:
Seems less crashy when we remove some complexity

Honeybunch committed:
No more crashes

Honeybunch committed:
Time slicing mesh uploading

Honeybunch committed:
Fixing minor warning

Honeybunch committed:
Smoother loading by breaking up scene loading

Honeybunch committed:
Separating shader optimization flags from debug info flags

Honeybunch committed:
Moving the concept of a free list to its own module so I can reuse it

Honeybunch committed:
Reducing per-frame work that has to be done by the render object system

Honeybunch commented:
Ok this got a bit away from me and there are still some rough edges but it's mostly a huge improvement. Scenes now load asynchronously and a large part of the renderer's descriptor management systems have been re-written to account for that.
